src/plugins/AvatarPEPPlugin.ts

```diff
@@ -0,0 +1,245 @@
+            return true;
+         }
+
+         if (info.length > 0) {
+            let hash = $(info).attr('id');
+
+            let storedHash = this.getStorage().getItem(from.bare);
+            if (storedHash === undefined || hash !== storedHash) {
+               let avatarUI = AvatarUI.get(contact);
+               this.getStorage().setItem(contact.getJid().bare, hash);
+               avatarUI.reload();
+            }
+         } else {
+            let avatarUI = AvatarUI.get(contact);
+            this.getStorage().setItem(contact.getJid().bare, '');
+            avatarUI.reload();
+         }
+      } else if (data.length > 0) {
+         let contact = this.pluginAPI.getContact(from);
+         if (!contact) {
+            this.pluginAPI.Log.warn('No contact found for', from);
+            return true;
+         }
+
+         let src = data.text().replace(/[\t\r\n\f]/gi, '');
+         const b64str = src.replace(/^.+;base64,/, '');
+         let hash = Hash.SHA1FromBase64(b64str);
+         let avatarUI = AvatarUI.get(contact);
+         this.getStorage().setItem(contact.getJid().bare, hash);
+         avatarUI.reload();
+      }
+
+      return true;
+   };
+
+   private avatarProcessor = async (contact: IContact, avatar: IAvatar): Promise<[IContact, IAvatar]> => {
+      let storage = this.getStorage();
+      let hash = storage.getItem(contact.getJid().bare);
+
+      if (!hash && !avatar) {
+         try {
+            const avatarObject = await this.getAvatar(contact);
+            const data = avatarObject.src.replace(/^.+;base64,/, '');
+            avatar = new Avatar(Hash.SHA1FromBase64(data), avatarObject.type, avatarObject.src);
+
+            this.getStorage().setItem(contact.getJid().bare, avatar.getHash() || '');
+
+            let avatarUI = AvatarUI.get(contact);
+            avatarUI.reload();
+         } catch (err) {
+            // we could not find any avatar
+         }
+      }
+
+      if (!hash || avatar) {
+         return [contact, avatar];
+      }
+
+      try {
+         avatar = new Avatar(hash);
+      } catch (err) {
+         try {
+            const avatarObject = await this.getAvatar(contact);
+
+            if (avatarObject) {
+               avatar = new Avatar(hash, avatarObject.type, avatarObject.src);
+               return [contact, avatar];
+            } else {
+               this.pluginAPI.Log.warn('No local cached avatar found');
+               return [contact, avatar];
+            }
+         } catch (err) {
+            this.pluginAPI.Log.warn('Error during avatar retrieval', err);
+            return [contact, avatar];
+         }
+      }
+
+      return [contact, avatar];
+   };
+
+   private async getAvatar(contact: IContact): Promise<{ src: string; type: string }> {
+      let connection = this.pluginAPI.getConnection();
+
+      return connection
+         .getPEPService()
+         .retrieveItems(UANS_METADATA, contact.getJid().bare)
```

Comment:
I would add a `hash` parameter, because sometimes we have the hash already from the metadata notification.